File: GoogleAnalytics.Native/AnalyticsEngine.cpp
//
// AnalyticsEngine.cpp
// Implementation of the AnalyticsEngine class.
//

#include "pch.h"
#include "AnalyticsEngine.h"
#include "GAServiceManager.h"

using namespace GoogleAnalytics;
using namespace Platform;
using namespace Windows::Storage;

String^ AnalyticsEngine::Key_AppOptOut = "GoogleAnaltyics.AppOptOut";

AnalyticsEngine^ AnalyticsEngine::current = nullptr;

void AnalyticsEngine::SendPayload(Payload^ payload)
{
	if (!AppOptOut)
	{
		GAServiceManager::Current->SendPayload(payload);
	}
}

AnalyticsEngine::AnalyticsEngine(PlatformInfoProvider^ platformTrackingInfo)
{
	this->platformTrackingInfo = platformTrackingInfo;
}

AnalyticsEngine^ AnalyticsEngine::Current::get()
{
	if (current == nullptr)
	{
		current = ref new AnalyticsEngine(ref new PlatformInfoProvider());
	}
	return current;
}

Tracker^ AnalyticsEngine::GetTracker(String^ propertyId)
{
	if (propertyId == nullptr) propertyId = "";
	if (trackers.find(propertyId) == end(trackers))
	{
		auto tracker = ref new Tracker(propertyId, platformTrackingInfo, this);
		trackers[propertyId] = tracker;
		if (DefaultTracker == nullptr)
		{
			DefaultTracker = tracker;
		}
		return tracker;
	}
	else
	{
		return trackers[propertyId];
	}
}

void AnalyticsEngine::CloseTracker(Tracker^ tracker)
{
	trackers.erase(tracker->TrackingId);
	if (DefaultTracker == tracker)
	{
		DefaultTracker = nullptr;
	}
}
bool AnalyticsEngine::GetAppOptOut()
{
	if (ApplicationData::Current->LocalSettings->Values->HasKey(Key_AppOptOut))
	{
		appOptOut = (bool)ApplicationData::Current->LocalSettings->Values->Lookup(Key_AppOptOut);
	}
	else
	{
		appOptOut = false;
	}
	return appOptOut->Value;
}

bool AnalyticsEngine::AppOptOut::get()
{
	if (appOptOut != nullptr) return appOptOut->Value;
	return GetAppOptOut();
}
void AnalyticsEngine::AppOptOut::set(bool value)
{
	if (appOptOut == nullptr) GetAppOptOut();
	if (appOptOut->Value != value)
	{
		appOptOut = value;
		ApplicationData::Current->LocalSettings->Values->Insert(Key_AppOptOut, value);
		if (value) GAServiceManager::Current->Clear();
	}
}

File: GoogleAnalytics.Managed/Win80/AnalyticsEngine.cs
using GoogleAnalytics.Core;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Windows.Foundation;
using Windows.Storage;
using Windows.UI.Popups;

namespace GoogleAnalytics
{
    public class AnalyticsEngine
    {
        const string Key_AppOptOut = "GoogleAnaltyics.AppOptOut";

        TrackerManager manager;

        static AnalyticsEngine current;

        public static AnalyticsEngine Current
        {
            get
            {
                if (current == null)
                {
                    current = new AnalyticsEngine();
                }
                return current;
            }
        }

        private AnalyticsEngine()
        {
            manager = new TrackerManager(new PlatformInfoProvider());
        }

        bool isAppOptOutSet;
        public bool AppOptOut
        {
            get
            {
                if (!isAppOptOutSet) LoadAppOptOut();
                return manager.AppOptOut;
            }
            set
            {
                manager.AppOptOut = value;
                isAppOptOutSet = true;
                ApplicationData.Current.LocalSettings.Values[Key_AppOptOut] = value;
                if (value) GAServiceManager.Current.Clear();
            }
        }

        private void LoadAppOptOut()
        {
            if (ApplicationData.Current.LocalSettings.Values.ContainsKey(Key_AppOptOut))
            {
                manager.AppOptOut = (bool)ApplicationData.Current.LocalSettings.Values[Key_AppOptOut];
            }
            else
            {
                manager.AppOptOut = false;
            }
            isAppOptOutSet = true;
        }

        public IAsyncOperation<bool> RequestAppOptOutAsync()
        {
            return _RequestAppOptOutAsync().AsAsyncOperation();
        }

        async Task<bool> _RequestAppOptOutAsync()
        {
            MessageDialog msgDialog = new MessageDialog("Allow anonomous information to be collected to help improve this application?", "Help Improve User Experience");
            var optInCommand = new UICommand("Yes");
            var optOutCommand = new UICommand("No");
            msgDialog.Commands.Add(optInCommand);
            msgDialog.Commands.Add(optOutCommand);
            var dialogResult = await msgDialog.ShowAsync();
            var result = (dialogResult != optInCommand);
            AppOptOut = result;
            return result;
        }

        public bool IsDebugEnabled
        {
            get { return manager.IsDebugEnabled; }
            set { manager.IsDebugEnabled = value; }
        }

        public Tracker GetTracker(string propertyId)
        {
            return manager.GetTracker(propertyId);
        }

        public void CloseTracker(Tracker tracker)
        {
            manager.CloseTracker(tracker);
        }

        public Tracker DefaultTracker
        {
            get { return manager.DefaultTracker; }
            set { manager.DefaultTracker = value; }
        }
    }
}
